File: src/types/domain/journal.ts
import type { VerseSource } from './bible';

// Spiritual moment types
export type MomentType =
  | 'journal'
  | 'prayer'
  | 'devotional'
  | 'gratitude'
  | 'confession'
  | 'memory_practice'
  | 'obedience_step'
  | 'lectio'
  | 'examen'
  | 'answered_prayer';

// Journal media types
export type MediaType = 'photo' | 'voice' | 'drawing';

export interface JournalMedia {
  id: string;
  type: MediaType;
  uri: string;
  duration?: number; // for voice notes (seconds)
  thumbnail?: string; // for photos/drawings
  createdAt: Date;
}

// AI-generated insights for journal entries
export interface JournalAIInsights {
  summary?: string;
  suggestedVerses?: VerseSource[];
  reflectionQuestions?: string[];
  growthPatterns?: string[];
  generatedAt?: Date;
}

// Entry source context
export type JournalSourceType = 'standalone' | 'verse_reflection' | 'devotional' | 'ai_prompt' | 'bible_reading';

export interface JournalSource {
  type: JournalSourceType;
  referenceId?: string; // e.g., devotional day ID, verse reference
}

export interface SpiritualMoment {
  id: string;
  userId: string;
  momentType: MomentType;
  content: string;
  aiReflection?: string;
  linkedVerses: VerseSource[];
  sentimentScore?: number;
  themes: string[];
  createdAt: Date;
  updatedAt?: Date;
  metadata?: Record<string, unknown>;
  // Rich media support
  media?: JournalMedia[];
  // AI-generated insights
  aiInsights?: JournalAIInsights;
  // Entry status
  status?: 'draft' | 'published';
  // Entry source context
  source?: JournalSource;
}

// Journal draft for auto-save
export interface JournalDraft {
  id: string;
  content: string;
  linkedVerses: VerseSource[];
  media: JournalMedia[];
  lastSavedAt: Date;
  source?: JournalSource;
}

// AI prompt suggestion for journaling
export type JournalPromptType = 'morning' | 'evening' | 'verse_based' | 'theme_based' | 'contextual';

export interface JournalPrompt {
  id: string;
  type: JournalPromptType;
  text: string;
  relatedVerse?: VerseSource;
  icon: string;
}

// Obedience step tracking
export interface ObedienceStep {
  id: string;
  userId: string;
  sourceMomentId?: string;
  commitment: string;
  dueDate?: Date;
  completed: boolean;
  completedAt?: Date;
  followUpSent: boolean;
  reflection?: string;
  createdAt: Date;
}

// Timeline item for Journey screen
export interface TimelineItem {
  id: string;
  type: MomentType | 'answered_prayer' | 'memory_milestone';
  title: string;
  content: string;
  linkedVerses?: VerseSource[];
  themes?: string[];
  createdAt: Date;
}

// Seasonal rhythm
export type RhythmType = 'advent' | 'lent' | 'sabbath' | 'custom';

export interface RhythmEnrollment {
  id: string;
  userId: string;
  rhythmType: RhythmType;
  startDate: Date;
  endDate: Date;
  currentDay: number;
  completedDays: number[];
  createdAt: Date;
}

// Growth insights
export type InsightType = 'weekly' | 'monthly' | 'quarterly' | 'yearly' | 'theme';

export interface GrowthInsight {
  id: string;
  userId: string;
  periodStart: Date;
  periodEnd: Date;
  insightType: InsightType;
  title: string;
  narrative: string;
  keyMoments: string[];
  themesGrowth: Record<string, number>;
  createdAt: Date;
}

File: ios-native/ChooseGODApp/Core/Models/Journal.swift
import Foundation

// MARK: - Moment Type

enum MomentType: String, Codable, CaseIterable {
    case journal
    case prayer
    case devotional
    case gratitude
    case confession
    case memoryPractice = "memory_practice"
    case obedienceStep = "obedience_step"
    case lectio
    case examen
    case answeredPrayer = "answered_prayer"
}

// MARK: - Media Type

enum JournalMediaType: String, Codable {
    case photo
    case voice
    case drawing
}

// MARK: - Journal Media

struct JournalMedia: Identifiable, Codable, Equatable {
    let id: String
    let type: JournalMediaType
    let uri: String
    var duration: TimeInterval?
    var thumbnail: String?
    let createdAt: Date
    
    enum CodingKeys: String, CodingKey {
        case id, type, uri, duration, thumbnail
        case createdAt = "created_at"
    }
}

// MARK: - AI Insights

struct JournalAIInsights: Codable, Equatable {
    var summary: String?
    var suggestedVerses: [VerseSource]?
    var reflectionQuestions: [String]?
    var growthPatterns: [String]?
    var generatedAt: Date?
    
    enum CodingKeys: String, CodingKey {
        case summary
        case suggestedVerses = "suggested_verses"
        case reflectionQuestions = "reflection_questions"
        case growthPatterns = "growth_patterns"
        case generatedAt = "generated_at"
    }
}

// VerseSource is defined in ChatModels.swift — shared across Journal and Chat

// MARK: - Journal Source

enum JournalSourceType: String, Codable {
    case standalone
    case verseReflection = "verse_reflection"
    case devotional
    case aiPrompt = "ai_prompt"
    case bibleReading = "bible_reading"
}

struct JournalSource: Codable, Equatable {
    let type: JournalSourceType
    var referenceId: String?
    
    enum CodingKeys: String, CodingKey {
        case type
        case referenceId = "reference_id"
    }
}

// MARK: - Spiritual Moment

struct SpiritualMoment: Identifiable, Codable, Equatable {
    let id: String
    let userId: String
    let momentType: MomentType
    var content: String
    var aiReflection: String?
    var linkedVerses: [VerseSource]
    var sentimentScore: Double?
    var themes: [String]
    let createdAt: Date
    var updatedAt: Date?
    var metadata: [String: String]?
    var media: [JournalMedia]?
    var aiInsights: JournalAIInsights?
    var status: JournalStatus?
    var source: JournalSource?
    
    enum CodingKeys: String, CodingKey {
        case id
        case userId = "user_id"
        case momentType = "moment_type"
        case content
        case aiReflection = "ai_reflection"
        case linkedVerses = "linked_verses"
        case sentimentScore = "sentiment_score"
        case themes
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case metadata, media
        case aiInsights = "ai_insights"
        case status, source
    }
    
    var wordCount: Int {
        content.split(separator: " ").count
    }
    
    var formattedDate: String {
        let formatter = DateFormatter()
        formatter.dateStyle = .long
        formatter.timeStyle = .short
        return formatter.string(from: createdAt)
    }
    
    var shortDate: String {
        let formatter = DateFormatter()
        formatter.dateFormat = "MMM d"
        return formatter.string(from: createdAt)
    }
}

// MARK: - Journal Status

enum JournalStatus: String, Codable {
    case draft
    case published
}

// MARK: - Journal Draft

struct JournalDraft: Identifiable, Codable, Equatable {
    let id: String
    var content: String
    var linkedVerses: [VerseSource]
    var media: [JournalMedia]
    var lastSavedAt: Date
    var source: JournalSource?
    
    enum CodingKeys: String, CodingKey {
        case id, content, media, source
        case linkedVerses = "linked_verses"
        case lastSavedAt = "last_saved_at"
    }
}

// MARK: - Journal Prompt

enum JournalPromptType: String, Codable {
    case morning, evening, verseBased, themeBased, contextual
}

struct JournalPrompt: Identifiable {
    let id: String
    let type: JournalPromptType
    let text: String
    let relatedVerse: VerseSource?
    let icon: String
}

// MARK: - Filter / Search

enum JournalFilter: String, CaseIterable {
    case all = "All"
    case journal = "Journal"
    case prayer = "Prayer"
    case gratitude = "Gratitude"
    case devotional = "Devotional"
    
    var momentType: MomentType? {
        switch self {
        case .all: return nil
        case .journal: return .journal
        case .prayer: return .prayer
        case .gratitude: return .gratitude
        case .devotional: return .devotional
        }
    }
}
